### ISSUE TITLE
Failed to decode custom "name" section @...; ignoring (Maximum call stack size exceeded).

### ISSUE THREAD

**Issue by ShayRubach** [rustwasm/wasm-bindgen#3101]
### Summary

I've been using `wasm-pack` for quite to convert my rust code into wasm. That went well until some point where it all started to fall apart. I can't put my finger on when and why it happened (first I though it was a cagro update side effect, but now I just don't get it), but all of a sudden my `wasm-pack` pkg started to produce a different wasm binary where the JS files remained to same as always. I haven't made changes in the Rust code since the time it all worked to the time it didnt (now).

To my knowledge and after reviewing a lot of possible solutions over the course of my investigation, I suspect it might be related to the underlying `wasm-bindgen` logic. Please feel free to tell me I'm wrong.

### Additional Details

I'm running `wasm-pack build --target bundler` and then importing the pkg using `webpack 5` (refer to as `webpack-stream` in the logs).

The error log shows:
```
Failed to decode custom "name" section @24293091; ignoring (Maximum call stack size exceeded).
[16:08:33] 'build_npm' errored after 2.26 s
[16:08:33] Error in plugin "webpack-stream"
Message:
    Module parse failed: Unexpected section: 0x2e
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
Error: Unexpected section: 0x2e
```

I've tried all of the solutions I could find, from the outdated to the dated (not so much) ones with no success.
Main refrence: https://github.com/rustwasm/wasm_game_of_life/issues/22

- Downgrading `wasm-bindgen` to a lower version to avoid a bug introduced in some conjunction with webpack version along the say 
- Using webpack's type `javascript/auto` in the module.rule config (removed the error but introduced another one)
- Using webpack's `noParse` rule for avoid parsing the wasm file (removed the error but introduced another one)
- Using a dedicated `wasm-loader` during webpack build - the loader is deprecated as webpack moved to version 5
- Random stuff I scratched off people in discord

Nothing helped. I can't understand what's wrong with the produced binary as the Rust code was never changed! 
We have a working binary at the moment but the Rust code can't be updated until the wasm issue is solved. 

Please advise, anything is welcomed!
Cheers.

**Comment by Liamolucko** [comment]
My guess is that the first `Failed to decode custom "name" section` error left the parser in an invalid state, causing the second error, but I don't know what could be causing the first one. Apparently it was caused by a stack overflow ('Maximum call stack size exceeded'), but I don't know why that would've happened.

Would you be able to send me the binary that causes this error? I should hopefully be able to debug this more from there.

**Comment by ShayRubach** [comment]
Thank for the advice @Liamolucko. As much as I would love to send the binary and have someone look at it, I'm not allowed to at the moment. Maybe you can suggest what should I be looking for when examining the binary?
I can tell you that when I run my binary (20mb~) via `wasm2wat` I get a 400mb~ output .wat file that it is cut on its end. Meaning Atom can't preview the end of this file, and neither does the web wasm2jet page.

Do you perhaps have any further information on the first error? 
I understand the second one is clearly related to an overflow, which I'm not sure why/where is coming from..

**Comment by Liamolucko** [comment]
The overflow error gets caught and logged here:

https://github.com/xtuc/webassemblyjs/blob/c8be65e6702ca4f24588b5aa9d21b2c8c7b62f39/packages/wasm-parser/src/decoder.js#L1841

If you modify that code to log the error's stack trace, you should hopefully be able to see what caused the overflow.

This patch should do it:

```patch
diff --git a/packages/wasm-parser/src/decoder.js b/packages/wasm-parser/src/decoder.js
index 0a15cbb9..d1ea8550 100644
--- a/packages/wasm-parser/src/decoder.js
+++ b/packages/wasm-parser/src/decoder.js
@@ -1840,6 +1840,7 @@ export function decode(ab: ArrayBuffer, opts: DecoderOpts): Program {
             console.warn(
               `Failed to decode custom "name" section @${offset}; ignoring (${e.message}).`
             );
+            console.log(e.stack);
 
             eatBytes(offset - (initialOffset + remainingBytes));
           }
```

**Comment by ShayRubach** [comment]
Good to know, thanks. However, that would require using `wasm-bindgen` separately to convert the Rust code into wasm right? I can't use `wsam-pack` as I do now. Correct me if I'm wrong (haven't tried anything else so far).

**Comment by Liamolucko** [comment]
No, that code is part of one of Webpack's dependencies (`@webassemblyjs/wasm-parser`). You should be able to manually make that change to the compiled code in `node_modules` without having to recompile everything. The compiled version of that file should show up at `node_modules/@webassemblyjs/wasm-parser/esm/decoder.js` and `node_modules/@webassemblyjs/wasm-parser/lib/decoder.js` (I'm not sure which one gets used, so change both). Then just Ctrl+F for the 'Failed to decode custom "name" section' error message and add a statement to log the error's stack.

I should have been more clear up front, sorry.

**Comment by ShayRubach** [comment]
Oh good to know!
 
I have a small update in that manner that led me to solve the issue.

I found an interesting collision in my env during yesterday's investigation.. Apparently I had `Binaryen` (a wasm toolchain) installed locally on my mac. This toolchain installed `wasm-opt` into my `/usr/bin/wasm-opt`. Then, when `wasm-pack build` was invoked to build the wasm pkg off my Rust lib, it utilized that same `/usr/bin/wasm-opt`. Snippet from `wasm-pack build`:
```
[INFO]: Installing wasm-bindgen...
[INFO]: found wasm-opt at "/usr/bin/wasm-opt"
[INFO]: Optimizing wasm binaries with `wasm-opt`...
```

Along that, I had `wasm-opt` explicitly and manually configured to `false` my `Cargo.toml` because of some error I ran into along the way. Snippet from `Cargo.toml`:
```toml

[package.metadata.wasm-pack.profile.release]
wasm-opt = false

```

This probably collided with the above somehow and resulted in that undefined behaviour! Because once I removed the `wasm-opt` configuration from my `Cargo.toml`, the error went away. 

On my CICD yaml, I just made sure to remove the configuration and install the `Binaryen` toolchain beforehand and it worked.

**p.s:** the "all of a sudden" voodoo scenario I mentioned in my original question where my entire build had broke was probably the point where I installed the `Binaryen` toolchain, as it affected the `wasm-pack` build process.

Thanks a lot for you insight @Liamolucko
